lib/posix/pthread.c

```diff
@@ -708,12 +708,16 @@ int pthread_once(pthread_once_t *once, void (*init_func)(void))
 {
 	__unused int ret;
 
+	if (init_func == NULL) {
+		return EINVAL;
+	}
+
 	ret = k_mutex_lock(&pthread_once_lock, K_FOREVER);
 	__ASSERT_NO_MSG(ret == 0);
 
-	if (once->is_initialized != 0 && once->init_executed == 0) {
```

Comment:
Sure seems like this could use an atomic instead of a shared global mutex.

Reply:
Possibly. Zephyr currently does not support anything aside from longs for atomic variables though, so we would be back to the same size as before (on 64-bit systems). Of course, globally, there would be 1 fewer mutex.

I wasn't really planning on a complete overhaul in this PR though.

Just trying to get Zephyr to behave like other systems in order to complete C++ SDK work.

Reply:
@keith-packard - I think an overhaul is imminent, now that I look at it. Newlib is the odd-duck in this case, and if we're moving to picolibc (preferred), it doesn't make sense to cater to newlib.

Probably only need to alias the `pthread_once_t` structure and add a build assert.

Maybe I'll recycle this PR.